design/app-to-app.md

```diff
@@ -0,0 +1,111 @@
+// https://github.com/microsoft/mixed-reality-extension-sdk/blob/master/packages/sdk/src/core/context.ts
+
+/**
+ * Registers a set of callbacks invoked at various points in the processing
+ * of a sharing message. Only one callback may be registered at a time
+ * for a given callback type.
+ * 
+ * The messageReceived callback must be defined, but the isMessageDesired
+ * callback is optional.
+ * 
+ * Returns true if the callback was successfully registered.
+ * @event
+ */
+public onSharingMessage(messageType: string, callbacks: SharingCallbacks): bool {
+    // [... Register listener ...]
+    return true; // If successful
+}
+
+/**
+ * Removes the onSharingMessage callback, if it is registered.
+ * 
+ * Returns true if the callback was successfully unregistered.
+ * @event
+ */
+public offSharingMessage(messageType: string, callbacks: SharingCallbacks): bool {
+    // [... Deregister listener ...]
+    return true; // If successful
+}
+```
+
+## Implementation Suggestions
+
+1. Modify `context` to expose callback registration mechanism
+2. Adjust `contextInternal` and perhaps `multiPeerAdapter` to message the desire for each MessageType callback to each client (and corresponding remove messages). For callbacks registered for all sessions or all users, this my be sent after connect
+3. Implement the "Photo" message type. Send and receive encoded photos via the existing websocket used by multiPeerAdapter. Future message types (e.g., streaming video) may need to establish related side-channels.
+
+### Sample data flow for photo sharing
+1. On connect (or on subsequent sharing message registration), MRE app sends a `SharingRegistration` json payload consisting of
+    * `messageType` - `string` (as Described in [API Design](#API-Design))
+2. MRE Client sends a `SharingMessage` json payload, consiting of the following fields (as Described in [API Design](#API-Design))
+    * `messageType` - `string` 
+    * `photoType` - `string`
+    * `dataSize` - `number`
+    * `transmissionMethod` - `string`, presently set to the value "websocketbinary", but may be extended, for example, to support side channel for future message types, or alternative trasnport for existing message types.
+3. MRE app invokes `SharingCallbacks.isMessageDesired` callback if present to determine whether the message is desired. If a `SharingCallback` is registered for the given `messageType` but no `isMessageDesired` is present, proceed as if the message IS desired. If no `SharingCallback` is registered for the given `messageType`, proceed as if the message is NOT desired.
+4. MRE app sends a `SharingMessageResponse` json payload, consiting of the following fields:
+    * `result` - `boolean` true if the message is desired, false otherwise.
+5. If the message was desired, the MRE host should transmit the bytes of the image immediately following the `SharingMessageResponse` in binary format as per https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/send
```

Comment:
I'm uncertain about sending large data on the websocket.  I would tend to assume that other messages cannot be sent until the blob is fully transferred, which may cause the app to become unresponsive during a large upload?

Reply:
@thespivey  agreed on the concern, but I don't think we're boxing ourselves in here -- not much changes if we, now or down the road (let me know if you'd like me to add this now), add something like "transmissionMethod" = "secondaryWebsocketBinary", and specify that the app should open should establish a side channel (and additional web socket connection) at this stage... a few more steps but not a dramatic change.

Reply:
I'll take a stab at this.

Reply:
Are you saying you would envision the client application needing to open a new socket on a different port to the MRE server app?

Reply:
Talked offline : decided on restful approach over side websocket